### ISSUE TITLE
Tracks (and maybe other instances) evaluate as False in Python, not that great?

### ISSUE THREAD

**Issue by gilou** [PixarAnimationStudios/OpenTimelineIO#1232]
## Feature Request

Proposed Change to existing Behavior: otio.schema.Track instances evaluate to False, maybe True would be better?

## Description

Objects in Python/OTIO (probably not limited to Track) evaluate to False. Generally, non-empty items tend to evaluates as True in Python, so it feels a bit weird there, and requires extra checking when returning an existing Track.

This code will show that track evaluates as False:
```
track = otio.schema.Track("Test Track")
if track:
    print("Track evaluates as True")
else:
    print("Track evaluates as False")

```

I'm not sure about what this will break, I assume people do explicit `== False ` when needed…

## Context

Simple example would be trying to define a track name as a sort of singleton, either returning an existing track in the timeline, or False if not found:

```
def find_track(timeline, trackname):
  for track in timeline.tracks:
    if track.name == trackname:
      return track
  return False

```

Because of the fact that track will evaluate to False, I can't do something like `track = find_track(tl, "test") `and just do `if track: xxx`, I need to check the result explicitly against False (`if track != False`), whereas if a proper Track() object would evaluate as True, it would look a lot nicer.

**Comment by reinecke** [comment]
Hi @gilou,

Thanks for the feedback.

The tl;dr is that this behavior is by design to make OTIO as pythonic as possible. This would be a breaking change to the API and would likely make it less pythonic. See below for a suggestion for how to handle the specific use case cited in a pythonic way.

# Explanation

OTIO's composition types (`Track` and `Stack`), as well as `SerializableCollection` behave as Python [container types](https://docs.python.org/3/reference/datamodel.html#emulating-container-types). This means that when the collections are empty, they test "falsy" and when they contain items, they test "truthy" (see the [Python data model](https://docs.python.org/3/reference/datamodel.html#object.__len__) guidelines on this).

We can show this behavior by expanding your example above:

```python
>>> track = otio.schema.Track("Test Track")
>>> if track:
...     print("Track evaluates as True")
... else:
...     print("Track evaluates as False")
...
Track evaluates as False
>>> # Add a clip to make the collection "truthy"
>>> track.append(otio.schema.Clip("Test Clip"))
>>> if track:
...     print("Track evaluates as True")
... else:
...     print("Track evaluates as False")
...
Track evaluates as True
```

Because the OTIO collection types implement `__len__`, this behavior is actually [built-in to the Python language](https://docs.python.org/3/reference/datamodel.html#object.__bool__).

Here is a dramatically simplified illustration of the behavior:

```python
>>> class Foo(object):
...     pass
...
>>> bool(Foo())
True
>>> class Bar(object):
...     def __len__(self):
...         return 0
...
>>> bool(Bar())
False
```

Changing this behavior would be both out of line with the [Python Data Model](https://docs.python.org/3/reference/datamodel.html) guidelines and a significant breaking change to the API.

# Suggestion

In the `find_track` example above, the function returns `false` when it is unable to find a result. A more common pattern in Python would be to return `None` when a result can't be found and then use `if track is not None: xxx` to test if a result was found.

This is [the Python way](https://en.wikipedia.org/wiki/Option_type#Names_and_definitions) of returning an [Option type](https://en.wikipedia.org/wiki/Option_type) where the result is either nothing (`None`) or something (the found `Track` instance, in this example).

One other thing to consider is that `find_track` may be better named `find_first_track` because there could be multiple tracks with the same name.

**Comment by gilou** [comment]
understood, I was not sure about the "actual" pythonic way there, but it makes sense… So a "True" track is in fact a Track with a length > 0 ? That seems reasonable, sorry about the noise ;)

**Comment by reinecke** [comment]
Yeah, that's pretty much it. This can be thought of in terms of "evaluates True" vs. "is True".

For example, this can feel a bit odd:
```python
>>> track = otio.schema.Track("Example", children=[otio.schema.Clip("Example clip")])
>>> track == True
False
>>> bool(track)
True
```

But, as it turns out, this behavior is consistent with Python's built-in types:
```python
>>> my_list = ["example"]
>>> my_list == True
False
>>> bool(my_list)
True
```

When using a non-boolean type with an `if` or `while`, you are implicitly converting that into a boolean value as shown above.

**Comment by gilou** [comment]
Yup, no mistake here, but I only tested my proposition on empty tracks, which is not the scenario I'm aiming at, so basically, the way it behaves is the way I need it to and indeed fully Pythonic! My bad, no problem here, and thanks for clarifying!
